src/core/tsi/ssl_transport_security.cc

```diff
@@ -1450,9 +1523,19 @@ static tsi_result ssl_handshaker_next(
   if (ssl_handshaker_get_result(impl) == TSI_HANDSHAKE_IN_PROGRESS) {
     *handshaker_result = nullptr;
   } else {
-    size_t unused_bytes_size = received_bytes_size - bytes_consumed;
-    const unsigned char* unused_bytes =
-        unused_bytes_size == 0 ? nullptr : received_bytes + bytes_consumed;
+    // In TLS 1.3, the ClientFinished or ServerFinished record may have
+    // (encrypted) application data appended to the end of the record. In TLS
+    // 1.2, this is explicitly disallowed by the RFC; application data will
+    // never be appended to a handshake record.
```

Comment:
This comment doesn't make sense. It's not possible for application data to be in the same record as handshake in any TLS version. A record has a single content type.

What behavior were you observing here / what assumption was gRPC making?

Reply:
Sorry the comment was not very clear - I've changed the comment and I will give a full explanation below.

What the comment was trying to convey is: in TLS 1.3, the client can send a flight of messages to the server that contains (a) a TLS handshake record containing the ClientFinished, and (b) TLS application data records.

When this occurred, the application data records were left in the "read" BIO, and the gRPC stack had no way of knowing that it had any data that needed to be decrypted. Instead, the code I added does the following:

(1) remove the application data records from the "read" BIO at the end of the handshake, and
(2) set the unused bytes equal to these application data records so that the gRPC stack knows there is something to be decrypted.

Reply:
Per your other comment, there may be other scenarios where application data is sent in a flight of messages with a handshake message. I singled out the case of ClientFinished + application data because that was the only case where any error seemed to occur, and I added the comment to try to explain why `unused_bytes` was being populated in this way.

Reply:
> What the comment was trying to convey is: in TLS 1.3, the client can send a flight of messages to the server that contains (a) a TLS handshake record containing the ClientFinished, and (b) TLS application data records.

This is true with TLS 1.2 as well. It's not new with TLS 1.3. Fundamentally, one of the client or server is going to speak last in the handshake. If the side which speaks last in the handshake also speaks first in the application protocol, it will send handshake data immediately followed by application data. When this happens, whether or not they end up in the same write() call on the sender (BoringSSL doesn't behave any differently on the sender side between TLS 1.2 and 1.3), they may come in at the same read() call on the receiver.

In TLS 1.3, the client speaks last. In TLS 1.2 full handshakes, the server speaks last. In TLS 1.2 resumption, the client speaks last. In TLS 1.2 False Start both sides effectively speak last in the handshake. TLS 1.3 0-RTT is extra fun. Also, in HTTP/2, both sides speak first in the application protocol because they need to send SETTINGS. Thus this scenario can happen all over the place.

> When this occurred, the application data records were left in the "read" BIO, and the gRPC stack had no way of knowing that it had any data that needed to be decrypted.

It sounds like this code had a bug, which equally applied to TLS 1.2 and TLS 1.3, but due to some quirk of testing, it only showed up on TLS 1.2 unit tests but not TLS 1.3? You cannot assume that the handshake will consume all data that you received in any version of TLS.

> (1) remove the application data records from the "read" BIO at the end of the handshake, and

Note this is only reasonable if the BIO is just a buffer rather than a socket. This is also implicitly assuming that BoringSSL hasn't overread the data and buffered it internally. (That's probably an okay assumption though. There actually used to be code to overread and we may restore it, but it was always off by default.)

Reply:
Thanks for explaining David, that was super helpful. I didn't realize this could also occur in TLS 1.2, and that this specific case could occur in the case of TLS 1.2 resumption. Looking back at the tests where this error occurred, they only ever do full handshakes, which explains why this error did not present itself when only TLS 1.2 was used.

In fact, the only e2e test involving TLS resumption is [h2_ssl_session_reuse_test.cc](https://github.com/grpc/grpc/blob/master/test/core/end2end/h2_ssl_session_reuse_test.cc).

Additionally, the BIO is indeed a buffer (built using `BIO_new_bio_pair`, and my understanding is that BoringSSL's `SSL_do_handshake` API does not overread. (Please correct me if that's not the case though.)

RE the comment itself: I propose removing any reference to TLS versions and simply writing

`When the handshake is complete, there may be records containing application data in |impl->ssl|'s read BIO. These bytes must be removed from the BIO and passed to the |ssl_tsi_handshaker_result| as unused bytes so that the gRPC stack knows that it has application data that must be decrypted.`

WDYT?

Reply:
Thanks @davidben and @matthewstevenson88 for the discussion. The ssl_transport_security code has been there for a long time, created before my time. It is true that there is only one e2e test related to TLS resumption. But if indeed there is a bug that ssl_transport_security does not process first application data correctly when application data are sent along with handshake bytes in TLS 1.2, it is strange that nobody reports this bug. 

@yang-g In gRPC streaming RPC, can server send the application data to client first (before client sending any data to server)? It looks like client needs to send some metadata to server first.

Reply:
@jiangtaoli2016 If I understand how TLS 1.2 resumption works (@davidben please keep me honest), then I don't **think** this bug would surface to the user, it would just cause TLS 1.2 resumption to be slower than it should be. The flow would be:

1. Client sends ClientHello + ticket to the server.
2. Server responds.
3. Client sends ClientFinished + application data to the server.
4. Server completes handshake but then the application data is sitting in `impl->ssl`'s read BIO without the server knowing it is there, so the server hangs.
5. Now, either (a) the client sends more application data to the server, or (b) the client pings the server. In either case, this adds to the server's read BIO and causes **everything** in the read BIO to be decrypted. From this point, the server can continue as normal.

If this is correct, then none of this would be exposed to the user, aside from perhaps a bit of latency. After logging all the calls to read/write, this is what occurs in the `h2_ssl_session_reuse_test.cc` currently (i.e. without this PR), and I'm assuming the same occurs for users.

Moving forward, I think that we should add a unit test that checks this case. This can be done by modifying [`ssl_tsi_test_do_handshake_session_cache`](https://github.com/grpc/grpc/blob/0f28314557fbe77e61e0e44f4159b774710f65b3/test/core/tsi/ssl_transport_security_test.cc#L707).

Reply:
I guess I don't really understand the way gRPC handles I/O here, so I'm not sure what are the implications of leaving things in the BIO. I would have expected TLS 1.3 and TLS 1.2 resumption to have the same behavior either way. Their I/O patterns are pretty similar.

Reply:
Agreed that TLS 1.3 and TLS 1.2 resumption should have the same behavior. The reason for this discussion was that gRPC e2e tests do not use resumption (with 1 exception), so this bug with TLS 1.2 resumption was not discovered until we turned on TLS 1.3 (which ran against all the e2e tests). 

At the moment, I'm writing a test that explicitly checks that the TLS 1.2 resumption behavior works as expected when application data is sent in the same flight of messages as the ClientFinished. Such checks already exist for TLS 1.3, but not for TLS 1.2 resumption. I should be able to add that test to this PR this afternoon.

@davidben @jiangtaoli2016 If I missed any concerns or confusion, please let me know. 

For context, here's a brief description of how gRPC's TLS stack handles I/O:
1. During the handshake, add everything that was received from the peer to read BIO.
2. At the end of the handshake, anything that is left unused should be added to the `unused_bytes` field of the handshake result struct.
3. The next thing that gRPC does is decrypt lies in the unused bytes field. It does this by adding the unused bytes to the read BIO and calling `SSL_read`.
4. Anything further that is received from the peer is added to the read BIO and decrypted by calling `SSL_read`.

If we don't populate the unused bytes from the read BIO, then we require something else to trigger the call to `SSL_read`. This is what currently (i.e. before this PR) happens for TLS 1.2 resumption, and it breaks the above model.

Reply:
I agree with Matt's description on gRPC TLS stack handles IO.
In gRPC, it is always the client that sends the first application data, confirmed with @yang-g. So server finishes TLS handshake last and sends along with application data is not a possible situation for gRPC.

Now back to TLS 1.2 resumption.
> 5. Now, either (a) the client sends more application data to the server, or (b) the client pings the server."

For unary RPC, once client sends a request, it will wait server for response. I don't think client will ping server right away if server does not respond.

@matthewstevenson88 Please let us know how the TLS 1.2 resumption test progresses on this particular case.

Reply:
In the most recent commit, I've (a) changed the unit tests so that TLS 1.2 resumption is tested when application data is sent in the same flight as the last client handshake record, and (b) amended the comment (that sparked this discussion) to explain why the unused bytes must be extracted.

Note that there was already a boolean variable in `ssl_transport_security_test.cc` that, when set to "true", sends application data along with the final handshake message. Before this PR, this [variable was set to "false"](https://github.com/grpc/grpc/blob/635ded1749f990ffe6be0ca403e4b255cf62742f/test/core/tsi/ssl_transport_security_test.cc#L424) for some reason, and this PR sets it to "true". If you set this variable to true in the current gRPC master, then the TLS 1.2 resumption test fails for exactly the reason we discussed earlier. I'm not sure why this variable was ever set to "false".

> For unary RPC, once client sends a request, it will wait server for response. I don't think client will ping server right away if server does not respond. Or a streaming RPC, it is normal for client to wait for server's data without sending request to server (except the initial metadata).

To elaborate on my previous point: I observed that `h2_ssl_session_reuse_test.cc` was not failing because the client was sending another flight of application data messages and, when the server received this next flight, the server read everything (i.e. the application data received immediately after the handshake, plus the application data in the next flight). My comment on the client ping was just a guess as to why this error was not visible to the user.

Reply:
Thanks Matt for the additional test and confirm.